### PR TITLE
Fix #1540: overloaded get and isDefined in option-less patmat

### LINKED ISSUE #1540: Overloaded `isDefined` or `get` breaks option-less pattern matching
Overloading `isDefined` or `get` as in the following example crashes dotc on this [assertion](https://github.com/lampepfl/dotty/blob/80cbe114cc5d641d66512c58402eca4ae69f9f2a/src/dotty/tools/dotc/transform/Erasure.scala#L335):

```scala
class Casey1(val a: Int) {
  def isDefined: Boolean = true
  def isDefined(x: Int): Boolean = ???
  def get: Int = a
  def get(x: Int): String = ???
}
object Casey1 { def unapply(a: Casey1) = a }

object Test {
  def main(args: Array[String]): Unit = {
    val c @ Casey1(x) = new Casey1(0)
    assert(x == c.get)
  }
}
```

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #1540 .

When `isDefined` and `get` are overloaded, make sure the translated code use the correct version of the methods.

### CODE REVIEW

File: src/dotty/tools/dotc/transform/PatternMatcher.scala [hunk 1]
```diff
@@ -240,17 +240,21 @@ class PatternMatcher extends MiniPhaseTransform with DenotTransformer {
...
 
-              Block(
-                List(ValDef(tmpSym, prev)),
-                // must be isEmpty and get as we don't control the target of the call (prev is an extractor call)
-                ifThenElseZero(
-                  ref(tmpSym).select(nme.isDefined),
-                  Block(List(ValDef(b.asTerm, prevValue)), next)
-                )
+            val tmpSym = freshSym(prev.pos, prev.tpe, "o")
+            val prevValue = ref(tmpSym).select(getDenot.namedType).ensureApplied
```
Comment (odersky) [inline src/dotty/tools/dotc/transform/PatternMatcher.scala]: I don't think that can work. getDenot has a type that's dependent on `prev`. But the denotation gets added to another prefix, `tmpSym`. You could try instead:

    ref(tmpSym).select(getDenot.symbol).ensureApplied

Comment (liufengyun) [inline src/dotty/tools/dotc/transform/PatternMatcher.scala]: Thanks for the tip @odersky !

File: src/dotty/tools/dotc/transform/PatternMatcher.scala [hunk 2]
```diff
@@ -240,17 +240,21 @@ class PatternMatcher extends MiniPhaseTransform with DenotTransformer {
...
-                  Block(List(ValDef(b.asTerm, prevValue)), next)
-                )
+            val tmpSym = freshSym(prev.pos, prev.tpe, "o")
+            val prevValue = ref(tmpSym).select(getDenot.symbol).ensureApplied
+
+            Block(
+              List(ValDef(tmpSym, prev)),
+              // must be isEmpty and get as we don't control the target of the call (prev is an extractor call)
+              ifThenElseZero(
+                ref(tmpSym).select(isDefinedDenot.namedType),
```
Comment (liufengyun) [inline src/dotty/tools/dotc/transform/PatternMatcher.scala]: Thanks for the feedback @odersky , I was wondering why the second succeeded without change. Now I added another test that fails my initial attempt.

**Comment by odersky** [comment]
LGTM, thanks!

File: src/dotty/tools/dotc/transform/PatternMatcher.scala [hunk 3]
```diff
@@ -240,17 +240,21 @@ class PatternMatcher extends MiniPhaseTransform with DenotTransformer {
           val isDefined = extractorMemberType(prev.tpe, nme.isDefined)
 
           if ((isDefined isRef defn.BooleanClass) && getTp.exists) {
-            val tmpSym = freshSym(prev.pos, prev.tpe, "o")
-            val prevValue = ref(tmpSym).select("get".toTermName).ensureApplied
+            // isDefined and get maybe overloaded
```
Comment (liufengyun) [inline src/dotty/tools/dotc/transform/PatternMatcher.scala]: Thanks  for the catch @smarter , it's fixed in the latest commit.

**Pull request merged**